pkg/catalogv2/system/system_test.go

```diff
@@ -378,3 +391,212 @@ func TestIsInstalledExactVersion(t *testing.T) {
+					},
+				}}},
+			},
+			expected: nil,
+		},
+		{
+			name: "Should return nil after the operation pod finishes successfully with exact version install",
+			input: testInput{
+				name:         "test-chart",
+				exactVersion: "102.0.0+up1.0.0",
+			},
+			mocks: testMocks{
+				indexOutput: &repo.IndexFile{
+					Entries: map[string]repo.ChartVersions{
+						"test-chart": []*repo.ChartVersion{
+							{Metadata: &chart.Metadata{Version: "102.0.0+up1.0.0"}},
+							{Metadata: &chart.Metadata{Version: "102.1.0+up1.0.0"}},
+						},
+					},
+				},
+				upgradeOutput: &catalog.Operation{
+					Status: catalog.OperationStatus{
+						PodName:      "install-operation",
+						PodNamespace: "cattle-system",
+					},
+				},
+				podGetOutput: &v1.Pod{Status: v1.PodStatus{ContainerStatuses: []v1.ContainerStatus{
+					{
+						Name:  "helm",
+						State: v1.ContainerState{Terminated: &v1.ContainerStateTerminated{ExitCode: 0}},
+					},
+				}}},
+			},
+			expected: nil,
+		},
+	}
+
+	for _, test := range testCases {
+		if test.skip {
+			continue
+		}
+		var versionMatcher interface{}
+		if test.input.exactVersion != "" {
+			versionMatcher = mock.MatchedBy(func(r io.Reader) bool {
+				upgradeArgs := &types2.ChartUpgradeAction{}
+				_ = json.NewDecoder(r).Decode(upgradeArgs)
+				return upgradeArgs.Charts[0].Version == test.input.exactVersion
+			})
+		} else {
+			versionMatcher = mock.Anything
+		}
+
+		contentMock, opsMock, podsMock, settingsMock, helmMock, clusterRepoMock :=
+			&mocks.ContentClient{}, &mocks.OperationClient{}, &mocks.PodClient{}, &mocks.SettingController{}, &mocks.HelmClient{}, &mocks.ClusterRepoController{}
+
+		contentMock.On("Index", "", "rancher-charts", true).Return(test.mocks.indexOutput, test.mocks.indexError)
+		helmMock.On("ListReleases", test.input.namespace, test.input.name, action.ListDeployed).Return(test.mocks.isInstalledReleasesOutput, test.mocks.isInstalledReleasesError)
+		helmMock.On("ListReleases", test.input.namespace, test.input.name, action.ListPendingInstall|action.ListPendingUpgrade|action.ListPendingRollback).Return(test.mocks.hasStatusOutput, test.mocks.hasStatusError)
+		opsMock.On("Upgrade", context.TODO(), installUser, "", "rancher-charts", versionMatcher, test.input.installImageOverride).Return(test.mocks.upgradeOutput, test.mocks.upgradeError)
+		if test.mocks.podGetOutput != nil || test.mocks.podGetError != nil {
+			podsMock.On("Get", test.mocks.upgradeOutput.Status.PodNamespace, test.mocks.upgradeOutput.Status.PodName, metav1.GetOptions{}).Return(test.mocks.podGetOutput, test.mocks.podGetError)
+		}
+		manager, _ := NewManager(context.TODO(), contentMock, opsMock, podsMock, settingsMock, clusterRepoMock, helmMock)
+		err := manager.install(test.input.namespace, test.input.name, test.input.minVersion, test.input.exactVersion, test.input.values, test.input.forceAdopt, test.input.installImageOverride)
+		asserts.Equal(test.expected, err, test.name)
```

Comment:
Typically we want to unit test exported functions only, but I can see that this might require a non-trivial refactor. I think the Manager struct may be responsible for too much to properly test all code path here in an isolated way. You don't have to change anything right now but if you are a maintainer for this code I think you consider refactoring this in the near future.

Reply:
I agree 100%, but we decided that it would take a lot more time to refactor this than to just write tests for the code as is, but breaking down the manager struct into smaller and easily testable pieces is something we talked about but is not currently a priority